Mathlib/LinearAlgebra/RootSystem/Base.lean

```diff
@@ -0,0 +1,208 @@
+/-
+Copyright (c) 2025 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import Mathlib.LinearAlgebra.RootSystem.Defs
+
+/-!
+# Bases for root pairings / systems
+
+This file contains a theory of bases for root pairings / systems.
+
+## Implementation details
+
+For reduced root pairings `RootSystem.Base` is equivalent to the usual definition appearing in the
+informal literature. However for non-reduced root pairings, it is more restrictive because it
+includes axioms on coroots as well as on roots. For example by
+`RootPairing.Base.eq_one_or_neg_one_of_mem_support_of_smul_mem` it is clear that the 1-dimensional
+root system `{-2, -1, 0, 1, 2} ⊆ ℝ` has no base.
+
+For infinite root systems, `RootSystem.Base` is usually not the right notion: linear independence
+is too strong.
+
+## Main definitions / results:
+ * `RootSystem.Base`: a base of a root pairing.
+
+## TODO
+
+ * Develop a theory of base / separation / positive roots for infinite systems which specialises to
+   the concept here for finite systems.
+
+-/
+
+noncomputable section
+
+open Function Set Submodule
+
+variable {ι R M N : Type*} [CommRing R] [AddCommGroup M] [Module R M] [AddCommGroup N] [Module R N]
+
+namespace RootPairing
+
+/-- A base of a root pairing.
+
+For reduced root pairings `RootSystem.Base` is equivalent to the usual definition appearing in the
+informal literature. However for non-reduced root pairings, it is more restrictive because it
+includes axioms on coroots as well as on roots. For example by
+`RootPairing.Base.eq_one_or_neg_one_of_mem_support_of_smul_mem` it is clear that the 1-dimensional
+root system `{-2, -1, 0, 1, 2} ⊆ ℝ` has no base. -/
+structure Base (P : RootPairing ι R M N) where
+  /-- The set of roots / coroots belonging to the base. -/
+  support : Set ι
+  linInd_root : LinearIndependent R fun i : support ↦ P.root i
+  linInd_coroot : LinearIndependent R fun i : support ↦ P.coroot i
+  root_mem_or_neg_mem (i : ι) : P.root i ∈ AddSubmonoid.closure (P.root '' support) ∨
+                              - P.root i ∈ AddSubmonoid.closure (P.root '' support)
```

Comment:
Am I missing something, or is this the same as using the `AddSubgroup.closure`?

Reply:
Yes, it is different. You get the union of the positive and negative half cones.
For example, try this with the standard basis vectors: Instead of getting Z^n you get the two octants where all coordinates have the same sign.